File: frontend/src/types/translation.ts
/**
 * Translation types — match the real backend schemas exactly.
 *
 * POST /api/expenses/{expense_id}/translate
 *   Request:  { target_language: 'en' | 'th' }
 *   Response: ExpenseTranslationResponse
 *
 * Note: The backend translates title, notes, and item names only.
 * Merchant name is NOT translated by the backend.
 */

export type TranslationLanguage = 'en' | 'th'

/** Request body sent to POST /api/expenses/{id}/translate */
export interface ExpenseTranslationRequest {
  target_language: TranslationLanguage
}

/** One translated item in the translation response */
export interface TranslatedExpenseItem {
  item_id: number
  original_name: string | null
  name_en: string | null
  name_th: string | null
  translated_name: string | null
}

/** Full response from POST /api/expenses/{id}/translate */
export interface ExpenseTranslationResponse {
  expense_id: number
  source_language: TranslationLanguage
  target_language: TranslationLanguage
  translated_title: string | null
  translated_notes: string | null
  items: TranslatedExpenseItem[]
  reused_existing_translation: boolean
}

File: backend/app/schemas/translation.py
"""Pydantic schemas for on-demand expense translation.

Only dynamic user data is translated here — expense titles, notes, and item names.
Fixed frontend labels (Dashboard, Save, Delete, etc.) are not handled by this endpoint.
"""

from typing import Literal

from pydantic import BaseModel, field_validator


class ExpenseTranslationRequest(BaseModel):
    """Request body for POST /api/expenses/{expense_id}/translate."""

    target_language: Literal["en", "th"]

    @field_validator("target_language", mode="before")
    @classmethod
    def normalise_language(cls, v: object) -> str:
        """Accept 'EN' and 'TH' in addition to lowercase."""
        if isinstance(v, str):
            return v.strip().lower()
        return v  # let Literal validation fail with a clear message


class TranslatedExpenseItem(BaseModel):
    """One translated item in the translation response."""

    item_id: int
    original_name: str | None = None
    name_en: str | None = None
    name_th: str | None = None
    translated_name: str | None = None


class ExpenseTranslationResponse(BaseModel):
    """Response body for POST /api/expenses/{expense_id}/translate."""

    expense_id: int
    source_language: Literal["en", "th"]
    target_language: Literal["en", "th"]
    translated_notes: str | None = None
    items: list[TranslatedExpenseItem] = []
    reused_existing_translation: bool = False


# ── Internal schemas for AI translation output ───────────────────────────

class GeminiTranslatedItem(BaseModel):
    """One item as returned by the AI in the translation response."""

    item_id: int
    translated_name: str | None = None


class GeminiTranslationResult(BaseModel):
    """Full AI translation output — validated before any database write.

    item_ids_expected must be supplied by the caller so the validator can
    confirm no extra or missing IDs are present.
    """

    translated_notes: str | None = None
    items: list[GeminiTranslatedItem] = []

    def validate_item_ids(self, expected_ids: list[int]) -> None:
        """Raise ValueError when returned item IDs do not match expected IDs exactly."""
        returned_ids = [item.item_id for item in self.items]
        extra = set(returned_ids) - set(expected_ids)
        if extra:
            raise ValueError(f"AI returned unexpected item IDs: {extra}")
        missing = set(expected_ids) - set(returned_ids)
        if missing:
            raise ValueError(f"AI did not return translations for item IDs: {missing}")
